### PR TITLE
Added favorite feature for songs

### PR DESCRIPTION
**Pull request by theTendo**
Added the ability to flag songs as favorites both from the song row (see gif) and from the music controller bar.

favorite_feature_demo

### CODE REVIEW

**Comment by AshotN** [comment]
Sorry for lag, will test and merge this soon. Am trying to get the Client in par with source-changes now.

File: client/Contexts/MusicContext.tsx [hunk 1]
```diff
@@ -110,6 +110,10 @@ export const MusicContextProvider: React.FC<MusicContextProps> = (props) => {
         _playSong(nextSong);
     }, [_playSong, songQueue, songQueueIndex, userQCount]);
 
+    const toggleFlag = useCallback(() => {
```
Comment (AshotN) [inline client/Contexts/MusicContext.tsx:113]: Is there any specific reason this is a `useCallback`?

File: client/components/MusicControl.tsx [hunk 2]
```diff
@@ -6,15 +6,25 @@ import Slider from '@material-ui/core/Slider';
...
+
+const MusicControl: React.FC = (props) => {
     const musicContext = useContext(MusicContext);
 
     const [ratingToggle, setRatingToggle] = useState(false);
 
     const [isSeeking, setIsSeeking] = useState(false);
     const [seekPosition, setSeekPosition] = useState(-1);
 
+    const [error, setError] = useState<Error | AmpacheError>(null);
```
Comment (AshotN) [inline client/components/MusicControl.tsx:26]: This isn't being used.

File: client/components/MusicControl.tsx [hunk 3]
```diff
@@ -6,15 +6,25 @@ import Slider from '@material-ui/core/Slider';
...
+import { AuthKey } from '~logic/Auth';
+import { toast } from 'react-toastify';
+import AmpacheError from '~logic/AmpacheError';
 
-const MusicControl: React.FC = () => {
+interface MusicControlProps {
+    authKey?: AuthKey;
+}
+
+const MusicControl: React.FC = (props) => {
```
Comment (AshotN) [inline client/components/MusicControl.tsx:18]: You declared the interface but didn't apply it.

This should be `const MusicControl: React.FC<MusicControlProps> = (props) => {`

File: client/components/MusicControl.tsx [hunk 4]
```diff
@@ -24,6 +34,29 @@ const MusicControl: React.FC = () => {
...
+                    return toast.success('Song added to favorites');
+                }
+                toast.success('Song removed from favorites');
+            })
+            .catch((err) => {
+                if (favorite) {
+                    toast.error(
+                        '😞 Something went wrong adding song to favorites.'
+                    );
+                } else {
```
Comment (AshotN) [inline client/components/MusicControl.tsx:51]: The `else` could be avoided by using a `return`

**Review by AshotN — CHANGES_REQUESTED** [review]
I'm not sure why but my server is returning an error when I try to flag a song. Your code has some changes that need to be made, but is overall about right. I need to get this project a little cleaned up and then we can come back to merging this.

Thank you!

File: client/Contexts/MusicContext.tsx (hunk 1 above)
Comment (theTendo) [reply client/Contexts/MusicContext.tsx]: not really, I removed it in the new commit

File: client/components/MusicControl.tsx (hunk 2 above)
Comment (theTendo) [reply client/components/MusicControl.tsx]: deleted as well

File: client/components/MusicControl.tsx (hunk 3 above)
Comment (theTendo) [reply client/components/MusicControl.tsx]: okay, I made the change

File: client/components/MusicControl.tsx (hunk 4 above)
Comment (theTendo) [reply client/components/MusicControl.tsx:49]: I copied this fragment from other components such as `SongList`

https://github.com/ampache/ampache/blob/f4749d26a3b24daaadc1e23f4d49ada2731ad7ef/client/components/SongList/index.tsx#L125-L129

Feel free to change my PR as you wish

**Comment by theTendo** [comment]
> I'm not sure why but my server is returning an error when I try to flag a song. Your code has some changes that need to be made, but is overall about right. I need to get this project a little cleaned up and then we can come back to merging this.
> 
> Thank you!

I made most of the changes you pointed to, thanks for the review!

I don't get any error when trying to flag a song, as mentioned in the Telegram group, it may be due to source changes.

**Comment by lachlan-00** [comment]
@usox has fixed that issue in source-changes now so should be all good for that part

File: client/components/MusicControl.tsx [hunk 5]
```diff
@@ -6,8 +6,16 @@ import Slider from '@material-ui/core/Slider';
...
 import CurrentPlayingArt from '~components/CurrentPlayingArt/';
 import Rating from '~components/Rating/';
+import { flagSong } from '~logic/Song';
+import { AuthKey } from '~logic/Auth';
+import { toast } from 'react-toastify';
+import AmpacheError from '~logic/AmpacheError';
 
-const MusicControl: React.FC = () => {
+interface MusicControlProps {
+    authKey?: AuthKey;
```
Comment (AshotN) [inline client/components/MusicControl.tsx:15]: I just realized, this shouldn't be optional.

Comment (theTendo) [reply client/components/MusicControl.tsx]: changed :)

**Comment by AshotN** [comment]
Thank you for your work!

**Review by AshotN — APPROVED** [review]

**Comment by AshotN** [comment]
I am really crap at Git, but I got this merged and I'll clean up whatever is needed from here. Thank you!

**Pull request merged**